src/java/org/apache/cassandra/index/sai/disk/v1/SkinnyPrimaryKeyMap.java

```diff
@@ -0,0 +1,185 @@
+ * </ul>
+ *
+ * While the {@link Factory} is threadsafe, individual instances of the {@link SkinnyPrimaryKeyMap}
+ * are not.
+ */
+@NotThreadSafe
+public class SkinnyPrimaryKeyMap implements PrimaryKeyMap
+{
+    @ThreadSafe
+    public static class Factory implements PrimaryKeyMap.Factory
+    {
+        protected final MetadataSource metadataSource;
+        protected final LongArray.Factory tokenReaderFactory;
+        protected final LongArray.Factory partitionReaderFactory;
+        protected final SortedTermsReader partitionKeyReader;
+        protected final IPartitioner partitioner;
+        protected final PrimaryKey.Factory primaryKeyFactory;
+
+        private FileHandle tokensFile;
+        private FileHandle partitionsFile;
+        private FileHandle partitionKeyBlockOffsetsFile;
+        private FileHandle partitionKeyBlocksFile;
+
+        public Factory(IndexDescriptor indexDescriptor, SSTableReader sstable)
+        {
+            try
+            {
+                this.metadataSource = MetadataSource.loadGroupMetadata(indexDescriptor);
+                NumericValuesMeta tokensMeta = new NumericValuesMeta(metadataSource.get(indexDescriptor.componentName(IndexComponent.TOKEN_VALUES)));
+                this.tokensFile = indexDescriptor.createPerSSTableFileHandle(IndexComponent.TOKEN_VALUES);
+                this.tokenReaderFactory = new BlockPackedReader(tokensFile, tokensMeta);
+                NumericValuesMeta partitionsMeta = new NumericValuesMeta(metadataSource.get(indexDescriptor.componentName(IndexComponent.PARTITION_SIZES)));
+                this.partitionsFile = indexDescriptor.createPerSSTableFileHandle(IndexComponent.PARTITION_SIZES);
+                this.partitionReaderFactory = new MonotonicBlockPackedReader(partitionsFile, partitionsMeta);
+                NumericValuesMeta partitionKeyBlockOffsetsMeta = new NumericValuesMeta(metadataSource.get(indexDescriptor.componentName(IndexComponent.PARTITION_KEY_BLOCK_OFFSETS)));
+                SortedTermsMeta partitionKeysMeta = new SortedTermsMeta(metadataSource.get(indexDescriptor.componentName(IndexComponent.PARTITION_KEY_BLOCKS)));
+                this.partitionKeyBlockOffsetsFile = indexDescriptor.createPerSSTableFileHandle(IndexComponent.PARTITION_KEY_BLOCK_OFFSETS);
+                this.partitionKeyBlocksFile = indexDescriptor.createPerSSTableFileHandle(IndexComponent.PARTITION_KEY_BLOCKS);
+                this.partitionKeyReader = new SortedTermsReader(partitionKeyBlocksFile, partitionKeyBlockOffsetsFile, partitionKeysMeta, partitionKeyBlockOffsetsMeta);
+                this.partitioner = sstable.metadata().partitioner;
+                this.primaryKeyFactory = indexDescriptor.primaryKeyFactory;
+            }
+            catch (Throwable t)
+            {
+                throw Throwables.unchecked(Throwables.close(t, Arrays.asList(tokensFile, partitionsFile, partitionKeyBlocksFile, partitionKeyBlockOffsetsFile)));
```

Comment:
It would be nice to make these 4 fields `final`. I think it'll compile if we do...

```
close();
throw Throwables.unchecked(t);
```

Is it essential to collect the suppressed exceptions, or is it fine to just log the warnings?

Reply:
It works fine if we move all the `FileHandle` creations out of the `try` block. That looks much nicer.

I don't see an issue with leaving the close/throw as is after that change.

Reply:
I think I have an elegant solution to this problem. I have added a `cleanup` mechanism to the `IndexDescriptor.createPer*FileHandle` methods that allow a cleanup to be done if the file handle creation fails. In this case, this allows us to pass `this::close` to the method, so any previously opened file handles that need closing will be closed.

I was going to leave it there but realised we had a similar issue in `PerColumnIndexFiles`, so I made the change for per-SSTable and per-index versions of this method.